soc/nordic/nrf54h/poweroff.c

```diff
@@ -0,0 +1,182 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/sys/poweroff.h>
+#include <zephyr/toolchain.h>
+#include <zephyr/pm/policy.h>
+#include <zephyr/arch/common/pm_s2ram.h>
+#include <hal/nrf_resetinfo.h>
+#include <hal/nrf_lrcconf.h>
+#include <hal/nrf_memconf.h>
+#include <zephyr/cache.h>
+
+#if defined(NRF_APPLICATION)
+#define RAMBLOCK_CONTROL_BIT_ICACHE 1
+#define RAMBLOCK_CONTROL_BIT_DCACHE 2
+#define RAMBLOCK_POWER_ID	    0
+#define RAMBLOCK_CONTROL_OFF	    0
+#elif defined(NRF_RADIOCORE)
+#define RAMBLOCK_CONTROL_BIT_ICACHE 2
+#define RAMBLOCK_CONTROL_BIT_DCACHE 3
+#define RAMBLOCK_POWER_ID	    0
+#define RAMBLOCK_CONTROL_OFF	    0
+#else
+#error "Unsupported domain."
+#endif
+
+static void suspend_common(void)
+{
+	if (IS_ENABLED(CONFIG_DCACHE)) {
+		/* Flush, disable and power down DCACHE */
+		sys_cache_data_flush_all();
+		sys_cache_data_disable();
+		nrf_memconf_ramblock_control_enable_set(NRF_MEMCONF, RAMBLOCK_POWER_ID,
+							RAMBLOCK_CONTROL_BIT_DCACHE, false);
+	}
+
+	if (IS_ENABLED(CONFIG_ICACHE)) {
+		/* Flush, disable and power down ICACHE */
+		sys_cache_instr_disable();
+		nrf_memconf_ramblock_control_enable_set(NRF_MEMCONF, RAMBLOCK_POWER_ID,
+							RAMBLOCK_CONTROL_BIT_ICACHE, false);
+	}
+
+	/* Disable retention */
+	nrf_lrcconf_retain_set(NRF_LRCCONF010, NRF_LRCCONF_POWER_DOMAIN_0, false);
+	nrf_lrcconf_poweron_force_set(NRF_LRCCONF010, NRF_LRCCONF_POWER_DOMAIN_0, false);
+
+	k_cpu_idle();
+}
+
+void z_sys_poweroff(void)
+{
+	nrf_resetinfo_resetreas_local_set(NRF_RESETINFO, 0);
+	nrf_resetinfo_restore_valid_set(NRF_RESETINFO, false);
+
+	nrf_lrcconf_retain_set(NRF_LRCCONF010, NRF_LRCCONF_POWER_MAIN, false);
+	nrf_lrcconf_poweron_force_set(NRF_LRCCONF010, NRF_LRCCONF_POWER_MAIN, false);
+
+	nrf_lrcconf_task_trigger(NRF_LRCCONF010, NRF_LRCCONF_TASK_SYSTEMOFFREADY);
+
+	suspend_common();
+
+	CODE_UNREACHABLE;
+}
+
+__weak void z_pm_soc_resume(void)
+{
+}
+
+/* Resume domain after local suspend to RAM. */
+void z_pm_sys_resume(void)
+{
+	if (IS_ENABLED(CONFIG_ICACHE)) {
+		/* Power up and re-enable ICACHE */
+		nrf_memconf_ramblock_control_enable_set(NRF_MEMCONF, RAMBLOCK_POWER_ID,
+							RAMBLOCK_CONTROL_BIT_ICACHE, true);
+		sys_cache_instr_enable();
+	}
+
+	if (IS_ENABLED(CONFIG_DCACHE)) {
+		/* Power up and re-enable DCACHE */
+		nrf_memconf_ramblock_control_enable_set(NRF_MEMCONF, RAMBLOCK_POWER_ID,
+							RAMBLOCK_CONTROL_BIT_DCACHE, true);
+		sys_cache_data_enable();
+	}
+
+	/* Re-enable domain retention. */
+	nrf_lrcconf_retain_set(NRF_LRCCONF010, NRF_LRCCONF_POWER_DOMAIN_0, true);
+	nrf_lrcconf_poweron_force_set(NRF_LRCCONF010, NRF_LRCCONF_POWER_MAIN, true);
+	nrf_lrcconf_poweron_force_set(NRF_LRCCONF010, NRF_LRCCONF_POWER_DOMAIN_0, true);
```

Comment:
The domains shall not be forced on here. I guess this is a workaround for the incomplete `k_cpu_idle()` implementation for nRF54H. I would at least add a TODO comment here, because when `k_cpu_idle()` will be implemented, we'll encounter surprises like unexpected forcing power domains ON.